notes/2018-01-22-packaging.md

```diff
@@ -0,0 +1,104 @@
+Packaging
+=========
+
+This documents explains how we want to package and distribute Radicle.
+
+We will create packages for Brew, Debian, Pacman, and (possibly Nix). This
+packages will be made available through Github Releases.
+
+Haskell binaries will be statically linked with `-O3` optimization.
+
+Files to package
+----------------
+
+The Radicle package contains the `rad` entry command, the `rad-xxx` subcommands
+and Radicle source files.
+
+The `rad` command will be installed into a `PATH` directory. The subcommands and
+Radicle source files are installed into distribution specific directories. For
+Debian and Pacman this is `/usr/lib/radicle/modules` for the Radicle source
+files and `/usr/lib/radicle/bin` for the subcommands.
+
+Locating packaged files
+-----------------------
+
+The binaries distributed with the package need to locate files on the system the
+package is installed on.
+
+* The `rad` entry command locates all installed subcommands to be able to list
+  them and dispatch to them.
+* Radicle apps (implemented as subcommands) need to locate Radicle source files.
+* If Radicle app frontends that are written in radicle the interpreter also needs
+  to locate Radicle source files to be able to import them.
+
+In addition to installed packages the location mechanism needs to work for
+development builds (`stack build`), local tests, and CI tests.
+
+To implement file location of subcommands we are going to use Cabal’s
+[`Paths_radicle`][cabal-paths] module. `getBinDir` will return the folder
+containing all subcommands.
+
+File location for Radicle source files will be implemented by reading the
+`RADPATH` environment variable
+
+For local development and testing this mechanisms work out of the box. For
+packages the `rad` command will be a script with the following content.
+
+```
+#!/bin/sh
+radicle_bin=/usr/lib/radicle/bin
+RADPATH=/usr/lib/radicle/modules
+exec $radicle_bin/rad "@$"
+```
+
+Setting `radicle_bin` will determine the return value of `getBinDir`. Setting
+`RADPATH` will enable other commands to pick up Radicle source files.
+
+[cabal-paths]: https://cabal.readthedocs.io/en/latest/developing-packages.html#accessing-data-files-from-package-code
+
+Dependencies
+------------
+
+Radicle depends on the IPFS daemon. Packages must either register the IPFS
+daemon as a dependency or distribute the `ipfs` binary with them.
+
+Services
+--------
+
+For Radicle to work two services need to be running: the Radicle daemon and the
+IPFS daemon. For Debian in Pacman we will provide a Systemd Unit file for the
+Radicle daemon as well as installation instructions that include extra steps to
+enable the Radicle daemon service.
+
+The Radicle daemon Systemd Unit has a dependency on the IPFS Systemd Unit. If we
+distribute IPFS with the package we also need to add a Systemd Unit file for
+IPFS.
+
+We will investigate whether Brew provides a similar service management system
```

Comment:
Yes, launchd. Essentially, one needs to inline a plist. See for example [postgres](https://github.com/Homebrew/homebrew-core/blob/master/Formula/postgresql.rb#L83)

Reply:
👍 Added it to the notes.